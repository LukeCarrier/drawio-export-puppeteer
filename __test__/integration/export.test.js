const fs = require('fs').promises;
const path = require('path');
const puppeteer = require('puppeteer');
const Jimp = require('jimp');

describe('export', () => {
  const EXPORT_URL = 'file://' + path.normalize(path.join(__dirname, '/../../export.html'));
  const FIXTURES = path.normalize(path.join(__dirname, '../fixtures'));

  let browser, page;

  beforeAll(async () => {
    browser = await puppeteer.launch();
  });
  afterAll(async () => await browser.close());

  beforeEach(async () => {
    page = await browser.newPage();

    await page.goto(EXPORT_URL, {
      waitUntil: 'networkidle0',
    });
  });
  afterEach(async () => await page.close());

  async function render(inputFilename, pageIndex, format) {
    const input = await fs.readFile(path.join(FIXTURES, inputFilename));

    await page.evaluate(renderArgs => {
      render(...renderArgs);
    }, [input.toString(), pageIndex, format]);
  }

  it('exports the first page of the flowchart fixture to JPEG', async () => {
    expect.assertions(1);

    await render('flowchart.drawio', 0, 'jpg');
    const screenshot = await page.screenshot();
    const converter = await Jimp.read(screenshot);
    const converted = await converter.getBufferAsync(Jimp.MIME_PNG);

    expect(converted).toMatchImageSnapshot();
  });

  it('exports the first page of the flowchart fixture to PNG', async () => {
    expect.assertions(1);

    await render('flowchart.drawio', 0, 'png');

    expect(await page.screenshot()).toMatchImageSnapshot();
  });

  it('exports the second page of the flowchart fixture to PNG', async () => {
    expect.assertions(1);

    await render('flowchart.drawio', 1, 'png');

    expect(await page.screenshot()).toMatchImageSnapshot();
  });
});
