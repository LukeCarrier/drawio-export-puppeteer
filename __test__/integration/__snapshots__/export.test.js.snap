// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`export exports the first page of the flowchart fixture to SVG 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><!DOCTYPE svg PUBLIC \\"-//W3C//DTD SVG 1.1//EN\\" \\"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" xmlns:xlink=\\"http://www.w3.org/1999/xlink\\" version=\\"1.1\\" width=\\"285px\\" height=\\"395px\\" viewBox=\\"-0.5 -0.5 285 395\\"><defs/><g><path d=\\"M 62 42 L 62 81.88\\" fill=\\"none\\" stroke=\\"#d6b656\\" stroke-miterlimit=\\"10\\" pointer-events=\\"stroke\\"/><path d=\\"M 62 90.88 L 57.5 81.88 L 66.5 81.88 Z\\" fill=\\"none\\" stroke=\\"#d6b656\\" stroke-miterlimit=\\"10\\" pointer-events=\\"all\\"/><rect x=\\"2\\" y=\\"2\\" width=\\"120\\" height=\\"40\\" rx=\\"6\\" ry=\\"6\\" fill=\\"#fff2cc\\" stroke=\\"#d6b656\\" pointer-events=\\"all\\"/><g transform=\\"translate(-0.5 -0.5)\\"><switch><foreignObject pointer-events=\\"none\\" width=\\"100%\\" height=\\"100%\\" requiredFeatures=\\"http://www.w3.org/TR/SVG11/feature#Extensibility\\" style=\\"overflow: visible; text-align: left;\\"><div xmlns=\\"http://www.w3.org/1999/xhtml\\" style=\\"display: flex; align-items: unsafe center; justify-content: unsafe center; width: 118px; height: 1px; padding-top: 22px; margin-left: 3px;\\"><div data-drawio-colors=\\"color: rgb(0, 0, 0); \\" style=\\"box-sizing: border-box; font-size: 0px; text-align: center;\\"><div style=\\"display: inline-block; font-size: 12px; font-family: Helvetica; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: all; white-space: normal; overflow-wrap: normal;\\">Lamp doesn't work</div></div></div></foreignObject><text x=\\"62\\" y=\\"26\\" fill=\\"rgb(0, 0, 0)\\" font-family=\\"Helvetica\\" font-size=\\"12px\\" text-anchor=\\"middle\\">Lamp doesn't work</text></switch></g><path d=\\"M 62 172 L 62 201.88\\" fill=\\"none\\" stroke=\\"#d6b656\\" stroke-miterlimit=\\"10\\" pointer-events=\\"stroke\\"/><path d=\\"M 62 210.88 L 57.5 201.88 L 66.5 201.88 Z\\" fill=\\"none\\" stroke=\\"#d6b656\\" stroke-miterlimit=\\"10\\" pointer-events=\\"all\\"/><g transform=\\"translate(-0.5 -0.5)\\"><switch><foreignObject pointer-events=\\"none\\" width=\\"100%\\" height=\\"100%\\" requiredFeatures=\\"http://www.w3.org/TR/SVG11/feature#Extensibility\\" style=\\"overflow: visible; text-align: left;\\"><div xmlns=\\"http://www.w3.org/1999/xhtml\\" style=\\"display: flex; align-items: unsafe center; justify-content: unsafe center; width: 1px; height: 1px; padding-top: 192px; margin-left: 82px;\\"><div data-drawio-colors=\\"color: rgb(0, 0, 0); \\" style=\\"box-sizing: border-box; font-size: 0px; text-align: center;\\"><div style=\\"display: inline-block; font-size: 11px; font-family: Helvetica; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: all; white-space: nowrap;\\">Yes</div></div></div></foreignObject><text x=\\"82\\" y=\\"195\\" fill=\\"rgb(0, 0, 0)\\" font-family=\\"Helvetica\\" font-size=\\"11px\\" text-anchor=\\"middle\\">Yes</text></switch></g><path d=\\"M 112 132 L 151.88 132\\" fill=\\"none\\" stroke=\\"#d6b656\\" stroke-miterlimit=\\"10\\" pointer-events=\\"stroke\\"/><path d=\\"M 160.88 132 L 151.88 136.5 L 151.88 127.5 Z\\" fill=\\"none\\" stroke=\\"#d6b656\\" stroke-miterlimit=\\"10\\" pointer-events=\\"all\\"/><g transform=\\"translate(-0.5 -0.5)\\"><switch><foreignObject pointer-events=\\"none\\" width=\\"100%\\" height=\\"100%\\" requiredFeatures=\\"http://www.w3.org/TR/SVG11/feature#Extensibility\\" style=\\"overflow: visible; text-align: left;\\"><div xmlns=\\"http://www.w3.org/1999/xhtml\\" style=\\"display: flex; align-items: unsafe center; justify-content: unsafe center; width: 1px; height: 1px; padding-top: 122px; margin-left: 137px;\\"><div data-drawio-colors=\\"color: rgb(0, 0, 0); \\" style=\\"box-sizing: border-box; font-size: 0px; text-align: center;\\"><div style=\\"display: inline-block; font-size: 11px; font-family: Helvetica; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: all; white-space: nowrap;\\">No</div></div></div></foreignObject><text x=\\"137\\" y=\\"125\\" fill=\\"rgb(0, 0, 0)\\" font-family=\\"Helvetica\\" font-size=\\"11px\\" text-anchor=\\"middle\\">No</text></switch></g><path d=\\"M 62 92 L 112 132 L 62 172 L 12 132 Z\\" fill=\\"#fff2cc\\" stroke=\\"#d6b656\\" stroke-miterlimit=\\"10\\" pointer-events=\\"all\\"/><g transform=\\"translate(-0.5 -0.5)\\"><switch><foreignObject pointer-events=\\"none\\" width=\\"100%\\" height=\\"100%\\" requiredFeatures=\\"http://www.w3.org/TR/SVG11/feature#Extensibility\\" style=\\"overflow: visible; text-align: left;\\"><div xmlns=\\"http://www.w3.org/1999/xhtml\\" style=\\"display: flex; align-items: unsafe center; justify-content: unsafe center; width: 90px; height: 1px; padding-top: 130px; margin-left: 17px;\\"><div data-drawio-colors=\\"color: rgb(0, 0, 0); \\" style=\\"box-sizing: border-box; font-size: 0px; text-align: center;\\"><div style=\\"display: inline-block; font-size: 12px; font-family: Helvetica; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: all; white-space: normal; overflow-wrap: normal;\\">Lamp<br />plugged in?</div></div></div></foreignObject><text x=\\"62\\" y=\\"134\\" fill=\\"rgb(0, 0, 0)\\" font-family=\\"Helvetica\\" font-size=\\"12px\\" text-anchor=\\"middle\\">Lamp...</text></switch></g><rect x=\\"162\\" y=\\"112\\" width=\\"120\\" height=\\"40\\" rx=\\"6\\" ry=\\"6\\" fill=\\"#fff2cc\\" stroke=\\"#d6b656\\" pointer-events=\\"all\\"/><g transform=\\"translate(-0.5 -0.5)\\"><switch><foreignObject pointer-events=\\"none\\" width=\\"100%\\" height=\\"100%\\" requiredFeatures=\\"http://www.w3.org/TR/SVG11/feature#Extensibility\\" style=\\"overflow: visible; text-align: left;\\"><div xmlns=\\"http://www.w3.org/1999/xhtml\\" style=\\"display: flex; align-items: unsafe center; justify-content: unsafe center; width: 118px; height: 1px; padding-top: 132px; margin-left: 163px;\\"><div data-drawio-colors=\\"color: rgb(0, 0, 0); \\" style=\\"box-sizing: border-box; font-size: 0px; text-align: center;\\"><div style=\\"display: inline-block; font-size: 12px; font-family: Helvetica; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: all; white-space: normal; overflow-wrap: normal;\\">Plug in lamp</div></div></div></foreignObject><text x=\\"222\\" y=\\"136\\" fill=\\"rgb(0, 0, 0)\\" font-family=\\"Helvetica\\" font-size=\\"12px\\" text-anchor=\\"middle\\">Plug in lamp</text></switch></g><path d=\\"M 62 292 L 62 341.88\\" fill=\\"none\\" stroke=\\"#d6b656\\" stroke-miterlimit=\\"10\\" pointer-events=\\"stroke\\"/><path d=\\"M 62 350.88 L 57.5 341.88 L 66.5 341.88 Z\\" fill=\\"none\\" stroke=\\"#d6b656\\" stroke-miterlimit=\\"10\\" pointer-events=\\"all\\"/><g transform=\\"translate(-0.5 -0.5)\\"><switch><foreignObject pointer-events=\\"none\\" width=\\"100%\\" height=\\"100%\\" requiredFeatures=\\"http://www.w3.org/TR/SVG11/feature#Extensibility\\" style=\\"overflow: visible; text-align: left;\\"><div xmlns=\\"http://www.w3.org/1999/xhtml\\" style=\\"display: flex; align-items: unsafe center; justify-content: unsafe center; width: 1px; height: 1px; padding-top: 332px; margin-left: 82px;\\"><div data-drawio-colors=\\"color: rgb(0, 0, 0); \\" style=\\"box-sizing: border-box; font-size: 0px; text-align: center;\\"><div style=\\"display: inline-block; font-size: 11px; font-family: Helvetica; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: all; white-space: nowrap;\\">No</div></div></div></foreignObject><text x=\\"82\\" y=\\"335\\" fill=\\"rgb(0, 0, 0)\\" font-family=\\"Helvetica\\" font-size=\\"11px\\" text-anchor=\\"middle\\">No</text></switch></g><path d=\\"M 112 252 L 151.88 252\\" fill=\\"none\\" stroke=\\"#d6b656\\" stroke-miterlimit=\\"10\\" pointer-events=\\"stroke\\"/><path d=\\"M 160.88 252 L 151.88 256.5 L 151.88 247.5 Z\\" fill=\\"none\\" stroke=\\"#d6b656\\" stroke-miterlimit=\\"10\\" pointer-events=\\"all\\"/><g transform=\\"translate(-0.5 -0.5)\\"><switch><foreignObject pointer-events=\\"none\\" width=\\"100%\\" height=\\"100%\\" requiredFeatures=\\"http://www.w3.org/TR/SVG11/feature#Extensibility\\" style=\\"overflow: visible; text-align: left;\\"><div xmlns=\\"http://www.w3.org/1999/xhtml\\" style=\\"display: flex; align-items: unsafe center; justify-content: unsafe center; width: 1px; height: 1px; padding-top: 242px; margin-left: 137px;\\"><div data-drawio-colors=\\"color: rgb(0, 0, 0); \\" style=\\"box-sizing: border-box; font-size: 0px; text-align: center;\\"><div style=\\"display: inline-block; font-size: 11px; font-family: Helvetica; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: all; white-space: nowrap;\\">Yes</div></div></div></foreignObject><text x=\\"137\\" y=\\"245\\" fill=\\"rgb(0, 0, 0)\\" font-family=\\"Helvetica\\" font-size=\\"11px\\" text-anchor=\\"middle\\">Yes</text></switch></g><path d=\\"M 62 212 L 112 252 L 62 292 L 12 252 Z\\" fill=\\"#fff2cc\\" stroke=\\"#d6b656\\" stroke-miterlimit=\\"10\\" pointer-events=\\"all\\"/><g transform=\\"translate(-0.5 -0.5)\\"><switch><foreignObject pointer-events=\\"none\\" width=\\"100%\\" height=\\"100%\\" requiredFeatures=\\"http://www.w3.org/TR/SVG11/feature#Extensibility\\" style=\\"overflow: visible; text-align: left;\\"><div xmlns=\\"http://www.w3.org/1999/xhtml\\" style=\\"display: flex; align-items: unsafe center; justify-content: unsafe center; width: 90px; height: 1px; padding-top: 250px; margin-left: 17px;\\"><div data-drawio-colors=\\"color: rgb(0, 0, 0); \\" style=\\"box-sizing: border-box; font-size: 0px; text-align: center;\\"><div style=\\"display: inline-block; font-size: 12px; font-family: Helvetica; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: all; white-space: normal; overflow-wrap: normal;\\">Bulb<br />burned out?</div></div></div></foreignObject><text x=\\"62\\" y=\\"254\\" fill=\\"rgb(0, 0, 0)\\" font-family=\\"Helvetica\\" font-size=\\"12px\\" text-anchor=\\"middle\\">Bulb...</text></switch></g><rect x=\\"2\\" y=\\"352\\" width=\\"120\\" height=\\"40\\" rx=\\"6\\" ry=\\"6\\" fill=\\"#fff2cc\\" stroke=\\"#d6b656\\" pointer-events=\\"all\\"/><g transform=\\"translate(-0.5 -0.5)\\"><switch><foreignObject pointer-events=\\"none\\" width=\\"100%\\" height=\\"100%\\" requiredFeatures=\\"http://www.w3.org/TR/SVG11/feature#Extensibility\\" style=\\"overflow: visible; text-align: left;\\"><div xmlns=\\"http://www.w3.org/1999/xhtml\\" style=\\"display: flex; align-items: unsafe center; justify-content: unsafe center; width: 118px; height: 1px; padding-top: 372px; margin-left: 3px;\\"><div data-drawio-colors=\\"color: rgb(0, 0, 0); \\" style=\\"box-sizing: border-box; font-size: 0px; text-align: center;\\"><div style=\\"display: inline-block; font-size: 12px; font-family: Helvetica; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: all; white-space: normal; overflow-wrap: normal;\\">Repair Lamp</div></div></div></foreignObject><text x=\\"62\\" y=\\"376\\" fill=\\"rgb(0, 0, 0)\\" font-family=\\"Helvetica\\" font-size=\\"12px\\" text-anchor=\\"middle\\">Repair Lamp</text></switch></g><rect x=\\"162\\" y=\\"232\\" width=\\"120\\" height=\\"40\\" rx=\\"6\\" ry=\\"6\\" fill=\\"#fff2cc\\" stroke=\\"#d6b656\\" pointer-events=\\"all\\"/><g transform=\\"translate(-0.5 -0.5)\\"><switch><foreignObject pointer-events=\\"none\\" width=\\"100%\\" height=\\"100%\\" requiredFeatures=\\"http://www.w3.org/TR/SVG11/feature#Extensibility\\" style=\\"overflow: visible; text-align: left;\\"><div xmlns=\\"http://www.w3.org/1999/xhtml\\" style=\\"display: flex; align-items: unsafe center; justify-content: unsafe center; width: 118px; height: 1px; padding-top: 252px; margin-left: 163px;\\"><div data-drawio-colors=\\"color: rgb(0, 0, 0); \\" style=\\"box-sizing: border-box; font-size: 0px; text-align: center;\\"><div style=\\"display: inline-block; font-size: 12px; font-family: Helvetica; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: all; white-space: normal; overflow-wrap: normal;\\">Replace Bulb</div></div></div></foreignObject><text x=\\"222\\" y=\\"256\\" fill=\\"rgb(0, 0, 0)\\" font-family=\\"Helvetica\\" font-size=\\"12px\\" text-anchor=\\"middle\\">Replace Bulb</text></switch></g></g><switch><g requiredFeatures=\\"http://www.w3.org/TR/SVG11/feature#Extensibility\\"/><a transform=\\"translate(0,-5)\\" xlink:href=\\"https://www.diagrams.net/doc/faq/svg-export-text-problems\\" target=\\"_blank\\"><text text-anchor=\\"middle\\" font-size=\\"10px\\" x=\\"50%\\" y=\\"100%\\">Viewer does not support full SVG 1.1</text></a></switch></svg>"`;
